### DISCUSSION TITLE
Pyinstaller seems to "execute" my files at build time

### DISCUSSION THREAD

**Discussion by igorbga** [pyinstaller/pyinstaller#8585] [Help]
I have a program that when running initializes some config information. This information is only available after runnning the program.
Some other modules of my program use that information and in case some of them tries to use the configuration before it is initialization it throws an exception. I have some modules that use that config just by the fact that they are imported.

We have noticed that during the build process of our ".exe" file the exception is thrown just after Pyinstaller prints "Looking for dynamic libraries".

Despite that exception Pyinstaller goes on and the ".exe" seems to be correctly generated.

But the thing is that this seems to imply that during build time Pyinstaller is in fact "running" the code in my modules, not just analyzing them. 

That was unexpected for us and could potentially be an issue depending on the sideeffects this execution could have.

Just in case it matters the package that triggers the exception is imported in a modules "\_\_init\_\_". Maybe is just the \_\_init\_\_ files that Pyinstaller does run?

Are we right to assume that our code will be, at least partially, run also at build time ? Should we organize our code taking this into consideration?

**Comment by rokm** [comment]
On Windows, we import packages in an isolated sub-process prior to running binary dependency analysis ("Looking for dynamic libraries"), in order to catch dynamic changes to DLL search paths that are made by some packages during their initialization. 

See: https://github.com/pyinstaller/pyinstaller/blob/3aa87b04648e5d9f957aec629fd2c0b283cb2bec/PyInstaller/building/build_main.py#L173-L273

**Reply by igorbga** [reply]
A ok, so I guess the answer is YES, PyInstaller does import packages during build time so any "permanent" sideeffect that the import might have should be considered.

I might have missed it but I think that I didn't find this mentioned anywhere in the documentation and it might be worth doing so as a warning.

By the way, is there any sensible way to know when we are running in build time instead of runtime? That would allow us to just to check it and avoid unwanted sideeffects during build time.

In worst case I guess we could just set an environment variable before launching PyInstaller in order to flag that we are running PyInstaller or something like that, but if there is already an "official" way to do so I'd rather not reinvent the wheel.

Thanks a lot for you fast and accurate answer.

**Reply by bwoodsend** [reply]
The *official* way to do it would be to not build importable modules with side-effects.

**Reply by rokm** [reply]
I suppose using environment variable would be most deliberate and thus also most foolproof.
